docs-next/pages/apis/hooks.mdx

```diff
@@ -2,6 +2,455 @@ import { Markdown } from '../../components/Page';
+| `listKey`       | The key of the list being operated on.                                                                                                                                            |
+| `fieldPath`     | The path of the field being operated on (field hooks only).                                                                                                                       |
+| `operation`     | The operation being performed (`'create'` or `'update'`).                                                                                                                         |
+| `originalInput` | The value of `data` passed into the mutation.                                                                                                                                     |
+| `existingItem`  | The currently stored item (`undefined` for `create` operations). This object is an unresolved list item. [list item API](./list-items) for more details on unresolved list items. |
+| `resolvedData`  | A [`resolved data`](#resolved-data-stages) object. The resolved data value after default values, relationship resolvers, and field resolvers have been applied.                   |
+| `context`       | The [`KeystoneContext`](./context) object of the originating GraphQL operation.                                                                                                   |
+
+```typescript
+import { config, createSchema, list } from '@keystone-next/keystone/schema';
+import { text } from '@keystone-next/fields';
+
+export default config({
+  lists: createSchema({
+    ListName: list({
+      hooks: {
+        validateInput: async ({
+          listKey,
+          operation,
+          originalInput,
+          existingItem,
+          resolvedData,
+          context,
+        }) => {
+          /* ... */
+          return resolvedData;
+        },
+      },
+      fields: {
+        fieldName: text({
+          hooks: {
+            validateInput: async ({
+              listKey,
+              fieldPath,
+              operation,
+              originalInput,
+              existingItem,
+              resolvedData,
+              context,
+            }) => {
+              /* ... */
+              return resolvedData[fieldName];
+            },
+          },
+        }),
+      },
+    }),
+  }),
+});
+```
+
+### validateInput
+
+The `validateInput` function is used to validate the [`resolvedData`](#resolved-data-stages) that will be saved during a `create` or `update` operation.
+
+It is invoked after the `resolveInput` hooks have been run.
+
+If the `resolvedData` is invalid then the function should report validation errors with `addValidationError(msg)` for list hooks, or `addFieldValidationError(msg)` for field hooks.
+These error messages will be returned as a `ValidationFailureError` from the GraphQL API.
+
```

Comment:
Should we clarify (and is it correct) that if any validation errors are on the stack after this part of the lifecycle, the API will return without calling further lifecycle methods or performing any changes to the database?

Reply:
Hmm, I would have thought that was obvious, but I can add a clarifying statement if you like.